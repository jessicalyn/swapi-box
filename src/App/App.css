.App {
  text-align: center;
}

button {
  margin: 2rem 3rem;
  padding: 1rem 2rem;
  border-radius: 1rem;
  font-family: 'Sawarabi Gothic', sans-serif;
  font-size: 1.5rem;
  background-color: #A5C4C9;
  color: #477480;
  border: .2rem solid #477480
}