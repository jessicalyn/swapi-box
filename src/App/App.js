import React, { Component } from 'react';
import './App.css';
import { fetchAnything } from '../Utils/fetchAnything'
import Buttons from '../Buttons/Buttons'
import { addHomeworldData, addSpeciesData } from '../Utils/simplifyPeople'
import { simplifyPlanets } from '../Utils/simplifyPlanets'
import { simplifyVehicles } from '../Utils/simplifyVehicles'
import CardContainer from '../Cards/CardContainer'

class App extends Component {
  constructor(){
    super();
    this.state = {
      error: "",
      movie: {},
      currentCards: []
    }
  }

  componentDidMount = async () => {
    const randomNumber = Math.floor((Math.random() * 7) + 1)
    const url = `https://swapi.co/api/films/${randomNumber}`
    const unfilteredMovie = await fetchAnything(url)
    this.showMovie(unfilteredMovie)
  }

  showMovie = (unfilteredMovie) => {
    this.setState({ movie: {
      title: unfilteredMovie.title,
      date: unfilteredMovie.release_date,
      crawl: unfilteredMovie.opening_crawl
    }})
  }

  showButtonCategory = async (category) => {
    if(category === "favorites"){
      this.setState({ error: "Favorites doesn't work yet, Jessica"})
    } else if(category === "people") {
      const unfilteredData = await fetchAnything(`https://swapi.co/api/${category}`)
      this.showPeople(unfilteredData)
    } else if(category === "planets") {
      const unfilteredData = await fetchAnything(`https://swapi.co/api/${category}`)
      this.showPlanets(unfilteredData)
    } else if(category === "vehicles") {
      const unfilteredData = await fetchAnything(`https://swapi.co/api/${category}`)
      this.showVehicles(unfilteredData)
    }
  }

  showPeople = async (unfilteredData) => {
    const addHomeworld = await addHomeworldData(unfilteredData)
    const simplifiedPeopleData = await addSpeciesData(addHomeworld)
    this.setState({ currentCards: simplifiedPeopleData })
  }

  showPlanets = async (unfilteredData) => {
    const simplifiedPlanetData = await simplifyPlanets(unfilteredData)
    this.setState({ currentCards: simplifiedPlanetData })
  }

  showVehicles = async (unfilteredData) => {
    const simplifiedVehicleData = await simplifyVehicles(unfilteredData)
    this.setState({ currentCards: simplifiedVehicleData })
  }

  render() {
    const { title, date, crawl } = this.state.movie
    const { error, currentCards } = this.state
    return (
      <div className="App">
        { error && <p>Error: { error }</p>}
        <h1>SwapiBox</h1>
        <Buttons showButtonCategory={ this.showButtonCategory }/>
        <CardContainer currentCards={ currentCards } />
        <div className="movieContainer">
          <h2>{ title }</h2>
          <h3>{ date }</h3>
          <h3>{ crawl }</h3>
        </div>
      </div>
    );
  }
}

export default App;
