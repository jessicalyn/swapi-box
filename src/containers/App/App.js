import React, { Component } from 'react';
import { connect } from 'react-redux'
import { Route } from 'react-router-dom'
import PropTypes from 'prop-types'
import './App.css';
import { fetchAnything } from '../../utils/fetchAnything'
import Buttons from '../../components/Buttons/Buttons'
import { addHomeworldData, addSpeciesData } from '../../utils/simplifyPeople'
import { simplifyPlanets } from '../../utils/simplifyPlanets'
import { simplifyVehicles } from '../../utils/simplifyVehicles'
import CardContainer from '../CardContainer/CardContainer'
import Movie from '../Movie/Movie'
import { addMovie, hasError, updateData, showCards, isLoading } from '../../actions'

export class App extends Component {

  componentDidMount = async () => {
    const randomNumber = Math.floor((Math.random() * 7) + 1)
    const url = `https://swapi.co/api/films/${randomNumber}`
    const unfilteredMovie = await fetchAnything(url)
    this.showMovie(unfilteredMovie)
  }

  showMovie = (unfilteredMovie) => {
    this.props.addMovie({
      title: unfilteredMovie.title,
      date: unfilteredMovie.release_date,
      crawl: unfilteredMovie.opening_crawl
    })
  }

  showButtonCategory = async (category) => {
    const { fetchedData } = this.props
    this.props.isLoading("true")
    const existingData = await fetchedData.filter(data => category === data.category)
    if(existingData.length > 0){
      this.props.isLoading("false")
      this.props.showCards(existingData)
    } else {
      this.fetchCategoryData(category)
    }
  }

  fetchCategoryData = async (category) => {
    if(category === "favorites"){
      this.props.hasError("Favorites doesn't work yet, Jessica")
    } else if(category === "people") {
      const unfilteredData = await fetchAnything(`https://swapi.co/api/${category}`)
      this.showPeople(unfilteredData)
    } else if(category === "planets") {
      const unfilteredData = await fetchAnything(`https://swapi.co/api/${category}`)
      this.showPlanets(unfilteredData)
    } else if(category === "vehicles") {
      const unfilteredData = await fetchAnything(`https://swapi.co/api/${category}`)
      this.showVehicles(unfilteredData)
    }
  }

  showPeople = async (unfilteredData) => {
    const addHomeworld = await addHomeworldData(unfilteredData)
    const simplifiedPeopleData = await addSpeciesData(addHomeworld)
    this.addFetchedData(simplifiedPeopleData)
    this.props.isLoading("false")
    this.props.showCards(simplifiedPeopleData)
  }

  showPlanets = async (unfilteredData) => {
    const simplifiedPlanetData = await simplifyPlanets(unfilteredData)
    this.addFetchedData(simplifiedPlanetData)
    this.props.isLoading("false")
    this.props.showCards(simplifiedPlanetData)
  }

  showVehicles = async (unfilteredData) => {
    const simplifiedVehicleData = await simplifyVehicles(unfilteredData)
    this.addFetchedData(simplifiedVehicleData)
    this.props.isLoading("false")
    this.props.showCards(simplifiedVehicleData)
  }

  addFetchedData = (categoryData) => {
    const { fetchedData } = this.props
    const addNewData = [...fetchedData, ...categoryData]
    this.props.updateData(addNewData)
  }

  render() {
    const { error, currentCards, isLoading, fetchedData } = this.props
    return (
      <div className="App">
        { error && <p>Error: { error }</p>}
        { isLoading === "true" && <h2>Your selection is loading</h2>}
        <h1>SwapiBox</h1>
        <Buttons showButtonCategory={ this.showButtonCategory }/>
        { fetchedData.length === 0 && <Movie /> }
        <CardContainer currentCards={ currentCards } />
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  error: state.error,
  isLoading: state.isLoading,
  movie: state.movie,
  fetchedData: state.fetchedData,
  currentCards: state.currentCards
})

const mapDispatchToProps = (dispatch) => ({
  addMovie: (movie) => dispatch(addMovie(movie)),
  hasError: (message) => dispatch(hasError(message)),
  updateData: (data) => dispatch(updateData(data)),
  showCards: (data) => dispatch(showCards(data)),
  isLoading: (message) => dispatch(isLoading(message))
})

export default connect(mapStateToProps, mapDispatchToProps)(App)
