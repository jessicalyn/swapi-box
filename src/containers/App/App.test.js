import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import { shallow } from 'enzyme'
// import { fetchAnything } from '../Utils/fetchAnything'
// import { addHomeworldData } from '../Utils/simplifyPeople'

const mockUnfiltered = {
  characters: [
    "https://swapi.co/api/people/1/", 
    "https://swapi.co/api/people/2/"
  ],
  created: "2014-12-18T10:39:33.255000Z",
  director: "Richard Marquand",
  edited: "2015-04-11T09:46:05.220365Z",
  episode_id: 6,
  opening_crawl: "Luke Skywalker has returned to his home planet of Tatooine in an attempt to rescue his friend Han Solo from the clutches of the vile gangster Jabba the Hutt. Little does Luke know that the GALACTIC EMPIRE has secretly begun construction on a new armored space station even more powerful than the first dreaded Death Star. When completed, this ultimate weapon will spell certain doom for the small band of rebels struggling to restore freedom to the galaxy...",
  planets: [
    "https://swapi.co/api/planets/5/", 
    "https://swapi.co/api/planets/7/", 
    "https://swapi.co/api/planets/8/", 
    "https://swapi.co/api/planets/9/", 
    "https://swapi.co/api/planets/1/"
  ],
  producer: "Howard G. Kazanjian, George Lucas, Rick McCallum",
  release_date: "1983-05-25",
  species: [
    "https://swapi.co/api/species/1/", 
    "https://swapi.co/api/species/2/"
  ],
  starships: [
    "https://swapi.co/api/starships/15/", 
    "https://swapi.co/api/starships/10/", 
    "https://swapi.co/api/starships/11/"
  ],
  title: "Return of the Jedi",
  url: "https://swapi.co/api/films/3/",
  vehicles: [
    "https://swapi.co/api/vehicles/8/", 
    "https://swapi.co/api/vehicles/16/", 
    "https://swapi.co/api/vehicles/18/"
  ]
}

const mockMovie = {
    crawl: "Luke Skywalker has returned to his home planet of Tatooine in an attempt to rescue his friend Han Solo from the clutches of the vile gangster Jabba the Hutt. Little does Luke know that the GALACTIC EMPIRE has secretly begun construction on a new armored space station even more powerful than the first dreaded Death Star. When completed, this ultimate weapon will spell certain doom for the small band of rebels struggling to restore freedom to the galaxy...",
    date: "1983-05-25",
    title: "Return of the Jedi"
};

const mockUrl = "https://swapi.co/api/people"

const mockError = "Something went wrong"

const mockLoading = "Your content is loading"

const mockFetchedData = [
  {
    category: "people",
    name: "Luke Skywalker"
  },
  {
    category: "people",
    name: "Chewbaca"
  },
  {
    category: "planet",
    name: "Earth"
  },
  {
    category: "vehicle",
    name: "Death Star"
  }
]

const mockCurrentCards = [
  {
    category: "people",
    name: "Luke Skywalker"
  },
  {
    category: "people",
    name: "Chewbaca"
  }
]

describe('App', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(
      <App 
        error={mockError} 
        isLoading={mockLoading}
        movie={mockMovie}
        fetchedData={mockFetchedData}
        currentCards={mockCurrentCards}
        />
    );
  });

  it('should match a snapshot with all data passed in', () => {
      expect(wrapper).toMatchSnapshot();
  });

})

