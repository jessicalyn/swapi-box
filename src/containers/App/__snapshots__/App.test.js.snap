// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match a snapshot with all data passed in 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(App)
    currentCards={
      Array [
        Object {
          "category": "people",
          "name": "Luke Skywalker",
        },
        Object {
          "category": "people",
          "name": "Chewbaca",
        },
      ]
    }
    error="Something went wrong"
    fetchedData={
      Array [
        Object {
          "category": "people",
          "name": "Luke Skywalker",
        },
        Object {
          "category": "people",
          "name": "Chewbaca",
        },
        Object {
          "category": "planet",
          "name": "Earth",
        },
        Object {
          "category": "vehicle",
          "name": "Death Star",
        },
      ]
    }
    isLoading="Your content is loading"
    movie={
      Object {
        "crawl": "Luke Skywalker has returned to his home planet of Tatooine in an attempt to rescue his friend Han Solo from the clutches of the vile gangster Jabba the Hutt. Little does Luke know that the GALACTIC EMPIRE has secretly begun construction on a new armored space station even more powerful than the first dreaded Death Star. When completed, this ultimate weapon will spell certain doom for the small band of rebels struggling to restore freedom to the galaxy...",
        "date": "1983-05-25",
        "title": "Return of the Jedi",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": [Function],
    },
    "ref": null,
    "rendered": [Function],
    "type": Object {
      "$$typeof": Symbol(react.context),
      "_calculateChangedBits": null,
      "_context": Object {
        "$$typeof": Symbol(react.context),
        "Consumer": [Circular],
        "Provider": Object {
          "$$typeof": Symbol(react.provider),
          "_context": [Circular],
        },
        "_calculateChangedBits": null,
        "_currentRenderer": null,
        "_currentRenderer2": null,
        "_currentValue": null,
        "_currentValue2": null,
        "_threadCount": 0,
      },
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": [Function],
      },
      "ref": null,
      "rendered": [Function],
      "type": Object {
        "$$typeof": Symbol(react.context),
        "_calculateChangedBits": null,
        "_context": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": [Circular],
          "Provider": Object {
            "$$typeof": Symbol(react.provider),
            "_context": [Circular],
          },
          "_calculateChangedBits": null,
          "_currentRenderer": null,
          "_currentRenderer2": null,
          "_currentValue": null,
          "_currentValue2": null,
          "_threadCount": 0,
        },
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
